asv/commands/run.py

```diff
@@ -151,31 +152,40 @@ def run(cls, conf, range_spec=None, steps=0, bench=None, parallel=1,
         elif range_spec == 'EXISTING':
             commit_hashes = [h for h, d in get_existing_hashes(
                 conf.results_dir)]
-        elif range_spec == 'NEW':
-            latest_result = find_latest_result_hash(
-                machine_params.machine, conf.results_dir)
-            commit_hashes = repo.get_hashes_from_range(
-                repo.get_new_range_spec(latest_result))
-        elif range_spec == "ALL":
-            commit_hashes = repo.get_hashes_from_range("")
+            commit_hashes = util.pick_n(commit_hashes, steps)
+        elif range_spec in ('NEW', 'ALL'):
+            branch_cache = BranchCache(conf, repo)
+            branch_commit_hashes = []
+            for branch in conf.branches:
+                if range_spec == 'NEW':
+                    branch_hashes = branch_cache.get_branch_commits(branch)
+                    latest_result = find_latest_result_hash(
+                        machine_params.machine, conf.results_dir,
+                        hashes=branch_hashes)
+                    spec = repo.get_new_range_spec(latest_result, branch)
+                else:
+                    spec = repo.get_branch_range_spec(branch)
+
+                new_hashes = repo.get_hashes_from_range(spec)
+                branch_commit_hashes.append(util.pick_n(new_hashes, steps))
+
+            # Splice hashes from branches
+            seen = set()
+            commit_hashes = []
+            for commit_hash in util.splice(*branch_commit_hashes):
+                if commit_hash not in seen:
+                    seen.add(commit_hash)
+                    commit_hashes.append(commit_hash)
```

Comment:
It looks like `steps` no longer applies when `NEW` or `ALL` is given...  I think steps should always apply (and probably just be handled below as it was before, rather than in select branches of the `if` statement here.